Source/WebCore/loader/cache/AllowedFonts.cpp

```diff
@@ -851,7 +851,10 @@ bool isFontBinaryAllowed(const void* data, size_t size, DownloadableBinaryFontAl
         return false;
     case DownloadableBinaryFontAllowedTypes::Restricted: {
         auto sha = hashForFontData(data, size);
-        return allowedFontHashesInLockdownMode().contains(sha);
+        auto allowedFontHashes = allowedFontHashesInLockdownMode().contains(sha);
+        if (!allowedFontHashes)
+            RELEASE_LOG(Fonts, "[Lockdown Mode] A font with a forbidden type has been blocked.");
```

Comment:
I think you'll need to include "Logging.h" to fix the build.

Reply:
I'd like to see if you could extract the font file name from `data` (or perhaps pass it from the containing function). having the name of the font file would be immensely helpful to debugging.

Reply:
@brentfulgham That seems problematic too, no? As if we do not log URI for privacy reasons, would it not be also an issue to log the name of the font face parsed from either the `src` or the `font-family` field .


```html
<style media="screen, print">
      @font-face {
        font-family: "MyPrivateFontOnIntranet";
        src: url("https://intranet.example.org/web-fonts/trackingCompanyInc.ttf");
      }
</style>
```

with cases also like these

```html
<style media="screen, print">
      @font-face {
        font-family: "MyPrivateFontOnIntranet";
        src: url("https://distributor.example.org/?id=345&version=italic");
      }
</style>
```

Or do we have access to the parsed font already. I though the current mechanism was blocking the download itself?

Reply:
Actually, an even better approach would be:
1. Log to WTFLogAlways that some undisclosed font was blocked due to LDM.
2. Log to Web Inspector console that a font with a specific URL was blocked by LDM.

That way, when investigating LDM site failures we could see in the log that LDM blocked a font.

So your current patch is fine as-is, but I'd love a follw-up that told the WebInspector console that we had blocked a specific font. Then developers (or I) could look at that and figure out what was going on.

Reply:
By `WTFLogAlways`, I assume you mean RELEASE_LOG() :)

I totally agree that we should log the failing URL when logging to the JS console. An earlier iteration was attempting to do this I believe though it was not working well. I had a proposal to fix it but Karl dropped it from the PR altogether.

Reply:
As discussed with @brentfulgham we could try to log the url of the font directly in the console log. So at least for the web developers or the person doing the diagnosis there will be a chance to understand what is breaking without leaking it in sysdiagnose.

Reply:
ok let me work on this today.

Reply:
> An earlier iteration was attempting to do this I believe though it was not working well. I had a proposal to fix it but Karl dropped it from the PR altogether.

yes. The message alone, a font has been blocked didn't seem useful in console.log. But if we add the URL which stays private to the user context, then it's entirely a different game. But this will not work in AllowedFonts.cpp 

Chris [suggested](https://github.com/WebKit/WebKit/pull/18684#discussion_r1348178273)

[`WebLocalFrameLoaderClient::dispatchDidFinishLoad()`](https://searchfox.org/wubkat/rev/e21786e1635702a6ca7c1e30290f32e8a97fec23/Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp#723) when `m_frame->isMainFrame()`

status: digging in the code.

Reply:
@cdumez hmm… dilemma…

[CachedFont::finishLoading](https://searchfox.org/wubkat/rev/e21786e1635702a6ca7c1e30290f32e8a97fec23/Source/WebCore/loader/cache/CachedFont.cpp#81-97) is the place where it is determined if the font should be loaded.

```cpp
if (!shouldAllowCustomFont(dataContiguous)) {
            setErrorAndDeleteData();
            return;
}
```

Above that we do not know if the font will be blocked (lockdown mode or others). 
Is the console accessible at this stage?

Reply:
It's a little bit of work but I think you can tweak this code here:
```
std::unique_ptr<FontLoadRequest> Document::fontLoadRequest(const String& url, bool isSVG, bool isInitiatingElementInUserAgentShadowTree, LoadedFromOpaqueSource loadedFromOpaqueSource)
{
    auto* cachedFont = m_fontLoader->cachedFont(completeURL(url), isSVG, isInitiatingElementInUserAgentShadowTree, loadedFromOpaqueSource);
    return cachedFont ? makeUnique<CachedFontLoadRequest>(*cachedFont) : nullptr;
}
```

to pass the Document to the CachedFontLoadRequest, like so:
```
std::unique_ptr<FontLoadRequest> Document::fontLoadRequest(const String& url, bool isSVG, bool isInitiatingElementInUserAgentShadowTree, LoadedFromOpaqueSource loadedFromOpaqueSource)
{
    auto* cachedFont = m_fontLoader->cachedFont(completeURL(url), isSVG, isInitiatingElementInUserAgentShadowTree, loadedFromOpaqueSource);
    return cachedFont ? makeUnique<CachedFontLoadRequest>(*cachedFont, *this) : nullptr;
}
```

Then you add a `Document&` parameter to the CachedFontLoadRequest constructor and save it as a data member:
```
WeakPtr<Document, WeakPtrImplWithEventTargetData> m_document;
```

Then in `CachedFont::finishLoading()`, set a boolean flag data member whenever we refuse to load the custom font:
e.g. `m_didRefuseToLoadCustomFont = true;`.

Also add a getter function for this flag.

Then in CachedFontLoadRequest::fontLoaded(), add logic like this:
```
if (m_font->didRefuseToLoadCustomFont() && m_document)
    m_document->addConsoleMessage(...);
```

Reply:
after wrestling with the errors and the help of @brentfulgham  this morning, A new version of the PR has been pushed. 

Thanks a lot to both of you. Learning a lot in the process. 

I fixed a couple more things since that screenshot. 
![Screenshot 2023-10-11 at 8 47 33](https://github.com/WebKit/WebKit/assets/505230/6c905493-6a7f-4606-a73a-76973a630970)